### PR TITLE
Add use_arrow as experimental feature to achieve better scalability

### PR DESCRIPTION
**Pull request by moomindani**
### Description

This PR is to add a new credential parameter `use_arrow` as experimental feature to achieve better scalability.

* Before this PR, each session writes result set into stdout and then it is passed to Interactive Session API response.
However, in larger use cases, it can cause `IllegalSessionStateException` due to its large result set size.

* This PR introduces a new credential parameter `use_arrow` (experimental, by default it's set to `False`)
  * When you set it to `True`, then dbt-glue writes result set into a feather file (an Arrow IPC file) and uploads to S3 bucket (`tmp-results` directory under `DBT_S3_LOCATION`). With this, each session passes only metadata (bucket and key location) without actual data. It reduces the amount of data passed over Interactive Session API.
  * This changes dbt-glue to call `GetSession` and `GetSecurityConfiguration` API to retrieve configured security configuration.
    * When SecurityConfiguration is set to the session, then dbt-glue refers it and follows its encryption mode to upload/download the feather file.

### Checklist

- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [ ] I have updated the `CHANGELOG.md` and added information about my change to the "dbt-glue next" section.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: dbt/adapters/glue/gluedbapi/connection.py [hunk 1]
```diff
@@ -382,7 +359,48 @@ def execute(cls,sql,output=True):
...
+            s3_client.upload_file(f"{session_id}.feather", result_bucket, result_key, ExtraArgs=extra_args)
+
+            # print and return only metadata instead of actual result data payload
+            del raw_results['results']
+            raw_results['result_bucket'] = result_bucket
+            raw_results['result_key'] = result_key
+            raw_results['description'] = description
+            dumped_results = json.dumps(raw_results, default=str)
+            print(dumped_results)
+            return dumped_results
```
Comment (aajisaka) [inline dbt/adapters/glue/gluedbapi/connection.py:401]: - Do we need these 3 lines?
- Do we need to call `print(dumped_results)` if `output` is false?

File: dbt/adapters/glue/gluedbapi/connection.py [hunk 2]
```diff
@@ -382,7 +359,48 @@ def execute(cls,sql,output=True):
...
+            s3_client = boto3.client('s3')
+            glue_client = boto3.client('glue')
+            security_config_name = glue_client.get_session(Id=session_id)["Session"].get("SecurityConfiguration", None)
+
+            o = urllib.parse.urlparse(location)
+            result_bucket = o.netloc
+            key = urllib.parse.unquote(o.path)[1:]
+            if not key.endswith('/'):
+                key = key + '/'
+            result_key = key + f"tmp-results/{session_id}.feather"
```
Comment (aajisaka) [inline dbt/adapters/glue/gluedbapi/connection.py:375]: If the user runs multiple dbt jobs concurrently with reusing the same session, is there any possibility to occur data race? If yes, we can add random ID to the key.

**Review by aajisaka — COMMENTED** [review]
Would you add a test case to run a dbt job with `use_arrow` feature enabled?

Given the feature is experimental, so I'm fine if the test case is added in a separate PR.

File: dbt/adapters/glue/gluedbapi/connection.py (hunk 1 above)
Comment (moomindani) [reply dbt/adapters/glue/gluedbapi/connection.py:401]: Yes definitely.
The param use_arrow=True is always used with output=True, and stdout is used to pass those values to Interactive Sessions API, and this function needs to return here. (Returned value won't used actually but I kept it for consistency with other place)

File: dbt/adapters/glue/gluedbapi/connection.py (hunk 2 above)
Comment (moomindani) [reply dbt/adapters/glue/gluedbapi/connection.py:375]: That's good catch, I will add the random ID as suffix.

Comment (moomindani) [reply dbt/adapters/glue/gluedbapi/connection.py]: Added

**Comment by moomindani** [comment]
> Would you add a test case to run a dbt job with `use_arrow` feature enabled?

Added the test case.

File: dbt/adapters/glue/gluedbapi/connection.py (hunk 1 above)
Comment (aajisaka) [reply dbt/adapters/glue/gluedbapi/connection.py:419]: Thanks. Make sense. Would you comment it to the source code?

Comment (moomindani) [reply dbt/adapters/glue/gluedbapi/connection.py:419]: Sure, added now.

**Comment by moomindani** [comment]
Closing this PR because I created another new one: https://github.com/aws-samples/dbt-glue/pull/323

**Pull request closed without merging**